### ISSUE TITLE
Exception when reflecting tables with SQL Server 2005

### ISSUE THREAD

**Issue by klapaukh** [sqlalchemy/sqlalchemy#5921]
**Describe the bug**

Reflecting a table's columns using a combination of the Microsoft ODBC SQL Server drivers with SQL Server 2005 raises an exception. 

**Expected behavior**
I would expect to get a list of columns in the table with no exception (which happens with the FreeTDS driver).

**To Reproduce**

```py
from sqlalchemy import create_engine, inspect

source_engine = create_engine("mssql+pyodbc:///?odbc_connect=[OMITTED]")

insp = inspect(source_engine)
insp.get_columns('TABLE NAME', schema='SCHEMA NAME')
```

**Error**

```
Traceback (most recent call last):
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
pyodbc.ProgrammingError: ('42000', "[42000] [Microsoft][ODBC SQL Server Driver][SQL Server]Incorrect syntax near 'sp_columns'. (102) (SQLExecDirectW); [42000] [Microsoft][ODBC SQL Server Driver][SQL Server]Statement(s) could not be prepared. (8180)")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\reflection.py", line 390, in get_columns
    col_defs = self.dialect.get_columns(
  File "<string>", line 2, in get_columns
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\reflection.py", line 52, in cache
    ret = fn(self, con, *args, **kw)
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\dialects\mssql\base.py", line 2195, in wrap
    return _switch_db(
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\dialects\mssql\base.py", line 2219, in _switch_db
    return fn(*arg, **kw)
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\dialects\mssql\base.py", line 2741, in get_columns
    cursor = connection.execute(
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\base.py", line 2235, in execute
    return connection.execute(statement, *multiparams, **params)
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\base.py", line 1011, in execute
    return meth(self, multiparams, params)
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\sql\elements.py", line 298, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\base.py", line 1124, in _execute_clauseelement
    ret = self._execute_context(
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\base.py", line 1316, in _execute_context
    self._handle_dbapi_exception(
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\base.py", line 1510, in _handle_dbapi_exception
    util.raise_(
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\util\compat.py", line 182, in raise_
    raise exception
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "C:\somewhere\miniconda3\envs\migration\lib\site-packages\sqlalchemy\engine\default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', "[42000] [Microsoft][ODBC SQL Server Driver][SQL Server]Incorrect syntax near 'sp_columns'. (102) (SQLExecDirectW); [42000] [Microsoft][ODBC SQL Server Driver][SQL Server]Statement(s) could not be prepared. (8180)")
[SQL: sp_columns @table_name = ?, @table_owner = ?]
[parameters: ('XXX', 'YYY')]
(Background on this error at: http://sqlalche.me/e/13/f405)
```

**Versions.**
 - OS: Windows 10 [& Debian 10 in docker image on a RedHat 7 host]
 - Python: 3.9
 - SQLAlchemy: 1.3.21
 - Database: Microsoft SQL Server 2005 - 9.00.1399.06 (Intel X86) 
 - DBAPI: pyodbc 4.0.30
 - Drivers: 
    - SQL Server 10.00.14393.00
    - SQL Server Native Client 11.0 2011.110.7462.06
    - ODBC Driver 17 for SQL Server 2017.174.01.01

**Additional context**
This bug doesn't happen with the FreeTDS Linux driver, or with never versions of SQL Server.

The way the expression is written matters. Using the implicit `EXEC` (as in the library) 
```py
rs = source_engine.execute("sp_columns @table_name = ?, @table_owner = ?", ('XXX', 'YYY'))
```
raises the exception seen above. However, using `EXEC` explicitly results in the correct behaviour (and should also be compatible with later SQL Server versions that do support the implicit form).
```py
rs = source_engine.execute("EXEC sp_columns @table_name = ?, @table_owner = ?", ('XXX', 'YYY'))
rs.fetchone()
```

**Have a nice day!**

**Comment by klapaukh** [comment]
This seems to have been fixed in 1.4.0

**Comment by zzzeek** [comment]
1.4 final is expected within a few weeks so it's not clear this is worth fixing.

**Comment by zzzeek** [comment]
OK can confirm EXEC doesn't break pymssql locally which would be my concern under 1.3.  However that call to sp_columns is really old so it's odd that nobody has ever reported this, is this speciifc to newer pyodbc versions + older SQL server?  cc @gordthompson

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_3** branch:

**Precede sp_columns with EXEC** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2535

**Comment by CaselIT** [comment]
@zzzeek has github changed how it closes the issues?

that's the second one that's not closed by the commit.

**Comment by zzzeek** [comment]
it says "fixes: #5921" right in the commit, i suspect github bug

**Comment by gordthompson** [comment]
For the record: Extended Support for SQL Server 2005 SP4 ended almost 5 years ago (April 2016). Last time I checked, most SQL Server tools, drivers, etc. only supported SQL Server 2008 and later, although that may have already changed since Extended Support for SQL Server 2008 ended over 18 months ago (July 2019).

**Comment by zzzeek** [comment]
https://github.blog/2013-01-22-closing-issues-via-commit-messages/

```

Now when you enter “Fixes #33” in a commit message, issue 33
will only be closed once the commit is merged into your
default branch (usually master).
```

was that other one also on a non-default branch?

**Comment by CaselIT** [comment]
don't remember, it may have been

**Comment by klapaukh** [comment]
Does SQL Alchemy 1.4.0 support SQL Server 2005 at all actually? 
I tried the beta which in get_indexes() tries to use the non-existent column `Sys.Indexes.filter_definition`.

I know that SQL Server 2005 & 2008 are old, but we still have databases running on them. And I am pretty sure that we are far from the only ones who do. 
I am currently running a locally hot fixed version of the 1.3 branch, so I'm probably fine for the short term. But it would be a bit tough I think if 1.3.x was the last time that SQL Server 2005 was supported.

**Comment by CaselIT** [comment]
I'm actually not sure. The [change log](https://docs.sqlalchemy.org/en/14/changelog/changelog_14.html#change-d336d859babeb66e0bdcba407bea5410) states that should support from 2005+ so that may be a bug.

The change that introduced the use of that column was https://github.com/sqlalchemy/sqlalchemy/commit/69502725db4829a84872697fd6569631d2a3c47f 

We don't currently have a mssql 2005 instance to test against, so if you could try also the other reflection methods in 1.4 and open an issue with the failing ones it would help us fix these errors.

As a sidenote, it there a location where one can check what an older version of mssql supports for a system tables? The documentation from microsoft https://docs.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-indexes-transact-sql has only from 2016

**Comment by klapaukh** [comment]
Ah, I was wondering if you had a test instance. 

I am happy to test the 1.4.0 reflection methods against the instance I'm using.

To be honest I am not familiar with SQL Alchemy, so I don't exactly know what comes under "other reflection methods". I've just been using the feature where you can use `Table(...)` to reflect what seems like everything. The inspector I just stumbled upon while debugging. Is there any chance you have a list (or script, or something) that just has "all reflection methods"? 

I have no idea where to find old documentation. If it's a problem, I could get the column lists at least from querying the running instance.

**Comment by gordthompson** [comment]
*(Re-opening to complete the discussion of whether 1.4 will continue to "support" SQL Server 2005.)*

Current master branch tested against SQL Server 2005 Standard:

pytest --db=mssql_198 C:/Users/Gord/git/sqla-gerrit/test/dialect/mssql/test_reflection.py

[test-dialect-mssql-test_reflection.txt](https://github.com/sqlalchemy/sqlalchemy/files/5946012/test-dialect-mssql-test_reflection.txt)

pytest --db=mssql_198 -k ComponentReflectionTest C:/Users/Gord/git/sqla-gerrit/test/dialect/test_suite.py

[suite-ComponentReflectionTest.txt](https://github.com/sqlalchemy/sqlalchemy/files/5946017/suite-ComponentReflectionTest.txt)

**Comment by CaselIT** [comment]
Thanks Gord.

It seems that most errors are the columns that's not found, but there are some other errors due to types (datetime2 and text that does not accept a size)

@gordthompson are you able to look into it?

**Comment by CaselIT** [comment]
It may be better to open a new issue, given the changelog etc, otherwise it may have a duplicate entry

**Comment by CaselIT** [comment]
Opened #5930 to track the 1.4 issue

**Comment by gordthompson** [comment]
For completeness, running the same two tests above with the current rel_1_3 branch (146efaf) against SQL Server 2005, the suite tests all passed and only two dialect tests failed:

https://github.com/sqlalchemy/sqlalchemy/blob/146efafcb436660e7891d3b34d05cd794c45268d/test/dialect/mssql/test_reflection.py#L35

failed because the duplicated table reported TEXT column length as `None` instead of `2_147_483_647`, and

https://github.com/sqlalchemy/sqlalchemy/blob/146efafcb436660e7891d3b34d05cd794c45268d/test/dialect/mssql/test_reflection.py#L324

fails because `TEXT(max)` is not valid in SQL Server 2005:

> sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Column, parameter, or variable #2: Cannot specify a column width on data type text. (2716) (SQLExecDirectW)')

Nothing terribly worrisome, IMO.
